src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/GeneratedDocumentSynchronizer.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Workspaces.Extensions;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal class GeneratedDocumentSynchronizer : DocumentProcessedListener
+    {
+        private readonly GeneratedDocumentPublisher _publisher;
+        private readonly DocumentVersionCache _documentVersionCache;
+        private readonly ProjectSnapshotManagerDispatcher _dispatcher;
+
+        public GeneratedDocumentSynchronizer(
+            GeneratedDocumentPublisher publisher,
+            DocumentVersionCache documentVersionCache,
+            ProjectSnapshotManagerDispatcher dispatcher)
+        {
+            _publisher = publisher;
+            _documentVersionCache = documentVersionCache;
+            _dispatcher = dispatcher;
+        }
+
+        public override void Initialize(ProjectSnapshotManager projectManager)
+        {
+        }
+
+        public override void DocumentProcessed(RazorCodeDocument codeDocument, DocumentSnapshot document)
+        {
+            _dispatcher.AssertDispatcherThread();
```

Comment:
Hmm, should we add an optional message to `AssertDispatcherThread` to give people a chance to point out which component is demanding the thread? Would improve our error messages in the case the assert fails and cause the demand to be self-documenting.

Reply:
I'd rather not only because the callstack should be sufficient in the dispatcher thread assert cases.